googlemock/include/gmock/gmock-spec-builders.h

```diff
@@ -985,7 +985,7 @@ class TypedExpectation<R(Args...)> : public ExpectationBase {
     return WillOnce(Action<F>(ActionAdaptor{
         std::make_shared<OnceAction<F>>(std::move(once_action)),
     }));
-  }
+  }  // NOLINT
```

Comment:
It's also possible to avoid the `NOLINT` with the following workaround:
```
     // Use a helper variable to avoid a clang-tidy false positive - see
     // https://github.com/llvm/llvm-project/issues/55219#issuecomment-1117665920
     auto action_ptr = std::make_shared<OnceAction<F>>(std::move(once_action));
     return WillOnce(Action<F>(ActionAdaptor{
       std::move(action_ptr),
     }));
```

Given that this is a know clang-tidy issue, I'm not sure whether this PR should even get merged to googletest. If it should, is there any preference between the explicit suppression and the workaround?

Reply:
?

Reply:
To add context, I'm not sure whether `// NOLINT` comments are the preferred approach to mitigate this in googletest, given https://github.com/google/googletest/pull/3676#issuecomment-1030228060 on a closed PR that would have added a bunch of NOLINT directives. The code I suggest is an alternative way to silence clang-tidy by avoiding the false-positive - it doesn't need the NOLINT when used. It comes from the suggested workaround in the LLVM issue tracker for this false positive.

My question is what is the preference from these options:
1. Use the NOLINT comment (merge this PR as-is)
2. Use the code mentioned in my previous comment, which avoids the false positive without the need for an explicit NOLINT. This avoids the possibly too-wide linter suppression, but adds some complexity.
3. None of the above, perhaps wait for clang-tidy to address the false-positive

Which option is preferred?

Reply:
NOLINT will surpress all warnings which is overly broad. I do not think we should be contorting our code to avoid false positives in a tool. I think the best course of action here is to file a bug with clang-tidy and wait for it to be fixed.

Reply:
Looking elsewhere in this repo, there's already a bunch of places where `// NOLINT` comments are used. Interestingly, a big part of these comes from the initial commit from 14 years ago. There is a bunch of PRs related to clang-tidy specific suppressions, most of them closed (typically due to inactivity). There is this one recent commit that made it in though: https://github.com/google/googletest/commit/4c5650f68866e3c2e60361d5c4c95c6f335fb64b. The difference I see on the merged one is that it's related to a true positive, rather than a false one.

Do I take it that in general, clang-tidy specific suppressions are accepted, as long as they are related to actual errors, but suppressions for false positives are rejected? I got mixed messages from the PR, as it was actually approved right away, but with question mark comments.

Reply:
We don't have official policy on this, and I'm sure our disposition has changed over time. I realize that's not a super helpful answer, but let me try to explain our current thinking:

GoogleTest is a project that gets included into many others and is therefore compiled and analyzed with nearly every configuration of warnings and clang-tidies. It's simply not possible for us to comply with every such request to avoid warning generation. If a warning uncovers an actual bug, or we can simultaneously improve the readability of our code and avoid the warning, we are willing to do so.

In this case in particular, the warning is incorrect and so adding suppression actually harms readability: Readers now need to understand why the code is correct and that there is some existing tool that is run by others in a configuration we do not directly test which incorrectly diagnoses the code as buggy. In other words, it would exist to circumvent a non-issue we do not test.